### PR TITLE
feat: auto-moderate the project channel

### LINKED ISSUE #413: [FEAT] - Automoderation of project-sharing
### What feature would you like to see?

👋 Curious what y'all think about adding this into the scope of the bot, but it'd be really useful to have something that automoderates the project-sharing channel in the Hacktoberfest Discord server.

Ideally, it would remove any message that doesn't contain a link to a GitHub/GitLab repository (or an issue within a repository), and DMs the user letting them know that their message was removed because of that.

The DM should explain that the channel is for sharing projects that are looking for help, and that for other conversations channels such as general should be used.

### Additional information

I'll need to grant the bot powers to delete messages in this channel if this feature does get approved. The bot should check it has that permission in the channel, and cleanly handle the situation if it doesn't.

### PR DESCRIPTION
**Pull request by naomi-lgbt**
# Pull Request

## Related Issue:

Closes #413

### CODE REVIEW

File: src/modules/checkProject.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  const hasProject =
+    /\bhttps?:\/\/(www\.)?git(hub|lab)\.com\/[^/]+\/[^/]+((\/-)?\/issues(\/\d+)?)?\/?\b/.test(
+      message.content
+    );
+
+  if (!hasProject) {
+    await message.delete();
+    await message.channel.send(
+      `<@!${message.author.id}>, please don't post in this channel if you aren't sharing a project.`
```
Comment (WalshyDev) [inline src/modules/checkProject.ts:48]: can we also make this message self-deleting? I feel like we'll end up with a bunch of these 😅

Comment (naomi-lgbt) [reply src/modules/checkProject.ts:48]: Hmmm - how long should it sit there?

Comment (MattIPv4) [reply src/modules/checkProject.ts:48]: Wondering if this'd be better as a DM?

Comment (naomi-lgbt) [reply src/modules/checkProject.ts:48]: My only concern is that Discord will quarantine our bot.

File: src/modules/checkProject.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    return;
+  }
+
+  if (matches.length > 1) {
+    await message.delete();
+    await message.channel.send(
+      `<@!${message.author.id}>, please only post one project at a time.`
+    );
+    return;
+  }
```
Comment (MattIPv4) [inline src/modules/checkProject.ts:59]: I don't think we need to worry about this for now? Worried we'd get false-positives with folks posting the top-level project link, and then some specific issues at the same time/

Comment (naomi-lgbt) [reply src/modules/checkProject.ts:59]: Blame @WalshyDev he wanted multi-link.

Maybe a check that `https://provider/owner/repo` is all the same?

Comment (WalshyDev) [reply src/modules/checkProject.ts:59]: I think going for a high number and tuning is good. I think posting 2 or 3 links is fine. Posting 10... that's a bit much

File: src/modules/checkProject.ts (hunk 1 above)
Comment (WalshyDev) [reply src/modules/checkProject.ts:48]: i think in channel with delete is fine. as for duration... not sure what a good time is but I think something like 45 seconds to a minute is more than enough.

File: src/modules/checkProject.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    !me ||
+    !me
+      .permissionsIn(channel as GuildChannelResolvable)
+      .has(PermissionFlagsBits.ManageMessages)
+  ) {
+    return;
+  }
+
+  const validLinkRegex =
+    /\bhttps?:\/\/(www\.)?git(hub\.com\/([!$&-;=-[\]_a-z~]|%[0-9a-fA-F]{2})+\/([!$&-;=-[\]_a-z~]|%[0-9a-fA-F]{2})+(\/issues(\/\d+))?|lab\.com(\/([!$&-;=-[\]_a-z~]|%[0-9a-fA-F]{2})+){2,}(\/-\/issues(\/\d+))?)\/?\b/gm;
```
Comment (MattIPv4) [inline src/modules/checkProject.ts:41]:
```suggestion
    /\bhttps?:\/\/(?:www\.)?git(?:hub\.com(\/[^\/#?\s]+\/[^\/#?\s]+)(?:\/issues(\/\d+))?|lab\.com((?:\/[^\/#?\s]+){2,})(?:\/-\/issues(\/\d+))?)\b/gmi;
```
Slightly less cursed regex (plus case-insensitive)

**Pull request merged**